### PR TITLE
ceph: configure health check timeout for OSDs when upgrading

### LINKED ISSUE #5790: Timeout during upgrade of OSDs needs to be configurable
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
There are a couple of different timeouts that are applied during an upgrade of OSDs that make it difficult to predict when the OSDs will be updated.

If `skipUpgrade: true`, the timeouts do not take effect.
If `continueUpgradeAfterChecksEvenIfNotHealthy: true`, the [5 minute timeout](https://github.com/rook/rook/blob/3ac06a533682eb5e40f954cf7255a9d28fb98ed7/pkg/operator/k8sutil/deployment.go#L81) will be skipped, but the latter [10 minute timeout](https://github.com/rook/rook/blob/8592caed58170683cbea7abec34bc37097635b83/pkg/daemon/ceph/client/upgrade.go#L221) will not be skipped.

These timeouts should be more consistent and also configurable.

The sequence is currently the following for the upgrade of each OSD daemon. 
1. Check for the OSD to be ok-to-stop. 
   a. If continueUpgradeAfterChecksEvenIfNotHealthy=true, skip to step 2 to update the osd immediately 
   b. If continueUpgradeAfterChecksEvenIfNotHealthy=false, wait up to 5 minutes for the OSD to be ok-to-stop. If the OSD is not ok-to-stop even after 5 minutes, skip the upgrade of this OSD and return to step 1 for the next OSD.
2. Update the OSD deployment
3. Check if it is ok-to-continue. Wait up to 10 minutes for the PGs to be clean
4. Repeat from step 1 for the next OSD

The same sequence applies to the update of the mds daemons, except step 5 has a shorter timeout of [2.5 minutes](https://github.com/rook/rook/blob/8592caed58170683cbea7abec34bc37097635b83/pkg/daemon/ceph/client/upgrade.go#L235). 

If skipUpgradeChecks=true, the above sequence does not apply since the OSD will always be updated immediately with no waits except for the pod to start after being updated.

**Expected behavior:**
With a new setting `waitTimeoutForHealthyOSDUpgrade` in the CephCluster CR, the flow would be improved to the following:

1. Check for the OSD to be `ok-to-stop`. Retry and wait up to the `waitTimeoutForHealthyOSDUpgrade`. 
   a. If `continueUpgradeAfterChecksEvenIfNotHealthy=true`, skip to step 2
   b. If `continueUpgradeAfterChecksEvenIfNotHealthy=false`, skip the upgrade of this OSD and return to step 1 for the next OSD.
2. Update the OSD deployment
3. Check if it is `ok-to-continue`. Retry and wait up to the `waitTimeoutForHealthyOSDUpgrade` for PGs to be clean.
4. Repeat from step 1 for the next OSD

The default timeout would be 5 minutes. Since this timeout may be hit both before and after the OSD is upgraded, the total timeout per OSD is actually double that setting.

**How to reproduce it (minimal and precise):**

- Install a cluster
- Upgrade the cluster while the PGs are unclean
- See each OSD delay 10-15 minutes before updated

### PR DESCRIPTION
**Pull request by sp98**
added WaitTimeoutForHealthyOSDUpgrade in the cluster crd that defines the time (in minutes) the operator would wait before an OSD can be stopped for
upgrade. It also defines the time the rook will wait for PGs to become active+clean after an OSD daemon is upgraded. The default value is 5 minutes.

Signed-off-by: Santosh Pillai <sapillai@redhat.com>

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #5790

Tested on multi-node local minikube cluster by upgrading from ceph 15.2.7 to 15.2.8

```Every 2.0s: kubectl -n rook-ceph get deployments -l rook_cluster=ro...  localhost.localdomain: Wed Jan 27 11:41:26 2021

rook-ceph-crashcollector-minikube-m02   req/upd/avl: 1/1/1      ceph-version=15.2.7-0
rook-ceph-crashcollector-minikube-m03   req/upd/avl: 1/1/1      ceph-version=15.2.7-0
rook-ceph-crashcollector-minikube-m04   req/upd/avl: 1/1/1      ceph-version=15.2.7-0
rook-ceph-mgr-a         req/upd/avl: 1/1/1      ceph-version=15.2.7-0
rook-ceph-mon-a         req/upd/avl: 1/1/1      ceph-version=15.2.7-0
rook-ceph-mon-b         req/upd/avl: 1/1/1      ceph-version=15.2.7-0
rook-ceph-mon-c         req/upd/avl: 1/1/1      ceph-version=15.2.7-0
rook-ceph-osd-0         req/upd/avl: 1/1/1      ceph-version=15.2.7-0
rook-ceph-osd-1         req/upd/avl: 1/1/1      ceph-version=15.2.7-0
rook-ceph-osd-2         req/upd/avl: 1/1/1      ceph-version=15.2.7-0
rook-ceph-osd-3         req/upd/avl: 1/1/1      ceph-version=15.2.7-0
rook-ceph-osd-4         req/upd/avl: 1/1/1      ceph-version=15.2.7-0
rook-ceph-osd-5         req/upd/avl: 1/1/1      ceph-version=15.2.7-0
```

```Every 2.0s: kubectl -n rook-ceph get deployments -l rook_cluster=ro...  localhost.localdomain: Wed Jan 27 11:45:16 2021

rook-ceph-crashcollector-minikube-m02   req/upd/avl: 1/1/1      ceph-version=15.2.8-0
rook-ceph-crashcollector-minikube-m03   req/upd/avl: 1/1/1      ceph-version=15.2.8-0
rook-ceph-crashcollector-minikube-m04   req/upd/avl: 1/1/1      ceph-version=15.2.8-0
rook-ceph-mgr-a         req/upd/avl: 1/1/1      ceph-version=15.2.8-0
rook-ceph-mon-a         req/upd/avl: 1/1/1      ceph-version=15.2.8-0
rook-ceph-mon-b         req/upd/avl: 1/1/1      ceph-version=15.2.8-0
rook-ceph-mon-c         req/upd/avl: 1/1/1      ceph-version=15.2.8-0
rook-ceph-osd-0         req/upd/avl: 1/1/1      ceph-version=15.2.8-0
rook-ceph-osd-1         req/upd/avl: 1/1/1      ceph-version=15.2.8-0
rook-ceph-osd-2         req/upd/avl: 1/1/1      ceph-version=15.2.8-0
rook-ceph-osd-3         req/upd/avl: 1/1/1      ceph-version=15.2.8-0
rook-ceph-osd-4         req/upd/avl: 1/1/1      ceph-version=15.2.8-0
rook-ceph-osd-5         req/upd/avl: 1/1/1      ceph-version=15.2.8-0
```

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

File: pkg/daemon/ceph/client/upgrade.go [hunk 1]
```diff
@@ -42,6 +42,9 @@ var (
 	// we don't perform any checks on these daemons
 	// they don't have any "ok-to-stop" command implemented
 	daemonNoCheck = []string{"mgr", "rgw", "rbd-mirror", "nfs"}
+
+	maxRetries = 5
+	delay      = 60 * time.Second
```
Comment (BlaineEXE) [inline pkg/daemon/ceph/client/upgrade.go:47]: This seems unused as set globally. Remove.

File: pkg/daemon/ceph/client/upgrade.go [hunk 2]
```diff
@@ -216,9 +224,9 @@ func okToStopDaemon(context *clusterd.Context, clusterInfo *ClusterInfo, deploym
 
 // okToContinueOSDDaemon determines whether it's fine to go to the next osd during an upgrade
 // This basically makes sure all the PGs have settled
-func okToContinueOSDDaemon(context *clusterd.Context, clusterInfo *ClusterInfo) error {
+func okToContinueOSDDaemon(context *clusterd.Context, clusterInfo *ClusterInfo, timeout int) error {
 	// Reconciliating PGs should not take too long so let's wait up to 10 minutes
-	err := util.Retry(10, 60*time.Second, func() error {
+	err := util.Retry(1, time.Duration(timeout)*time.Minute, func() error {
```
Comment (BlaineEXE) [inline pkg/daemon/ceph/client/upgrade.go:229]: Does this make it so the function no longer checks every minute for updates? Does this mean that OSDs aren't able to finish faster than the default 5 minutes?

File: pkg/daemon/ceph/client/upgrade.go [hunk 3]
```diff
@@ -160,12 +163,17 @@ func OkToStop(context *clusterd.Context, clusterInfo *ClusterInfo, deployment, d
 		if osdDoNothing(context, clusterInfo) {
 			return nil
 		}
+		maxRetries = 1
```
Comment (travisn) [inline pkg/daemon/ceph/client/upgrade.go:166]: With maxRetries=1, I believe we will sleep the whole time before checking again. Instead, we want a higher number of retries so we can check periodically before the whole timeout expired. For example, can we calculate a maxRetries that will cause a retry every 5-10s?

File: pkg/daemon/ceph/client/upgrade.go (hunk 2 above)
Comment (travisn) [reply pkg/daemon/ceph/client/upgrade.go:229]: I have the same concern above, I didn't see this comment soon enough. :)

File: cluster/charts/rook-ceph/templates/resources.yaml [hunk 4]
```diff
@@ -63,6 +63,8 @@ spec:
                   type: boolean
                 continueUpgradeAfterChecksEvenIfNotHealthy:
                   type: boolean
+                waitTimeoutForHealthyOSDUpgrade:
```
[the remark this answers is not in the document]
Comment (travisn) [reply cluster/charts/rook-ceph/templates/resources.yaml:66]: Or what if we make it like the `healthCheck` setting `interval`, which is a `string` such as `5m`?

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/cluster/mon/spec.go [hunk 5]
```diff
@@ -325,7 +325,7 @@ func (c *Cluster) makeMonDaemonContainer(monConfig *monConfig) v1.Container {
 }
 
 // UpdateCephDeploymentAndWait verifies a deployment can be stopped or continued
-func UpdateCephDeploymentAndWait(context *clusterd.Context, clusterInfo *client.ClusterInfo, deployment *apps.Deployment, daemonType, daemonName string, skipUpgradeChecks, continueUpgradeAfterChecksEvenIfNotHealthy bool) error {
+func UpdateCephDeploymentAndWait(context *clusterd.Context, clusterInfo *client.ClusterInfo, deployment *apps.Deployment, daemonType, daemonName string, skipUpgradeChecks, continueUpgradeAfterChecksEvenIfNotHealthy bool, healthyOSDTimeout int) error {
```
[the remark this answers is not in the document]
Comment (sp98) [reply pkg/operator/ceph/cluster/mon/spec.go:328]: This mon function `UpdateCephDeploymentAndWait` is used as `updateDeploymentAndWait` by all the daemons and also other operators like nfs. I've passed the OSD timeout as argument to this function.  I'm trying to figure out a better way to do this.

File: cluster/charts/rook-ceph/templates/resources.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (leseb) [reply cluster/charts/rook-ceph/templates/resources.yaml:66]: Also, I think `upgrade` is confusing, it's not only on upgrade by every time we restart an OSD, right? Every time the OSD deployment spec changes IIRC.

File: pkg/operator/ceph/cluster/mon/spec.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (sp98) [reply pkg/operator/ceph/cluster/mon/spec.go]: passing OSD timeout in ClusterInfo instance rather than an argument in `updateDeploymentAndWait` method.

File: pkg/daemon/ceph/client/upgrade.go (hunk 2 above)
Comment (sp98) [reply pkg/daemon/ceph/client/upgrade.go]: fixed.  Keeping the maxRetires value to 10 and retrying every `timeoutPeriodInSeconds/maxRetries`

File: cluster/examples/kubernetes/ceph/cluster.yaml [hunk 6]
```diff
@@ -37,6 +37,10 @@ spec:
   skipUpgradeChecks: false
   # Whether or not continue if PGs are not clean during an upgrade
   continueUpgradeAfterChecksEvenIfNotHealthy: false
+  # WaitTimeoutForHealthyOSD defines the time (in minutes) the operator would wait before an OSD can be stopped for upgrade or restart.
+  # It also defines the time the rook will wait for PGs to become active+clean after an OSD daemon is upgraded or restarted.
+  # The default wait timeout is 5 minutes.
+  # waitTimeoutForHealthyOSD: 10
```
Comment (travisn) [inline cluster/examples/kubernetes/ceph/cluster.yaml:43]: How about uncommenting and setting the default?
```suggestion
  waitTimeoutForHealthyOSD: 5
```

File: pkg/operator/ceph/cluster/osd/osd.go [hunk 7]
```diff
@@ -66,13 +67,14 @@ const (
...
 	osdAppNameFmt     = "rook-ceph-osd-%d"
 	// OsdIdLabelKey is the OSD label key
-	OsdIdLabelKey                  = "ceph-osd-id"
-	serviceAccountName             = "rook-ceph-osd"
-	portableKey                    = "portable"
-	cephOsdPodMinimumMemory uint64 = 2048 // minimum amount of memory in MB to run the pod
-	bluestorePVCMetadata           = "metadata"
-	bluestorePVCWal                = "wal"
-	bluestorePVCData               = "data"
+	OsdIdLabelKey                          = "ceph-osd-id"
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/osd.go:70]: Nothing changed in this block? Strange that the linter changed the spacing like this...

File: pkg/operator/ceph/cluster/osd/osd.go [hunk 8]
```diff
@@ -181,6 +183,13 @@ func (c *Cluster) Start() error {
 		logger.Warningf("useAllNodes is set to false and no nodes, driveGroups, storageClassDevicesets or volumeSources are specified, no OSD pods are going to be created")
 	}
 
+	if c.spec.WaitTimeoutForHealthyOSD != 0 {
+		c.clusterInfo.OsdUpgradeTimeout = (c.spec.WaitTimeoutForHealthyOSD * 60) * time.Second
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/osd.go:187]: Instead of setting a new `OsdUpgradeTimeout` on the clusterInfo, how about passing `c.spec` where necessary so the OSD upgrader can check this setting for itself? Then we could avoid the copied setting.

File: pkg/daemon/ceph/client/upgrade.go [hunk 9]
```diff
@@ -217,8 +227,9 @@ func okToStopDaemon(context *clusterd.Context, clusterInfo *ClusterInfo, deploym
 // okToContinueOSDDaemon determines whether it's fine to go to the next osd during an upgrade
 // This basically makes sure all the PGs have settled
 func okToContinueOSDDaemon(context *clusterd.Context, clusterInfo *ClusterInfo) error {
-	// Reconciliating PGs should not take too long so let's wait up to 10 minutes
-	err := util.Retry(10, 60*time.Second, func() error {
+	// for OSD, the delay between each retry should be waitTimeoutForHealthyOSD/maxRetries seconds
+	delay := clusterInfo.OsdUpgradeTimeout / maxRetries
```
Comment (travisn) [inline pkg/daemon/ceph/client/upgrade.go:231]: The timing is currently up to 1 minute for ok-to-stop, and 5 minutes (default) for ok-to-continue, right? This seems backwards to me. What about waiting 5 min for ok-to-stop and 1 min for ok-to-continue? If there are more OSDs to upgrade after this one, they will wait 5 min to upgrade with the ok-to-stop check.

File: pkg/daemon/ceph/client/upgrade.go [hunk 10]
```diff
@@ -27,6 +27,10 @@ import (
 	"github.com/rook/rook/pkg/util"
 )
 
+const (
+	maxRetries = 10
```
Comment (travisn) [inline pkg/daemon/ceph/client/upgrade.go:31]: If we have the same number of retries for different intervals, the individual retry interval will vary a lot. How about we calculate the number of retries based on retrying every 10s?

File: cluster/examples/kubernetes/ceph/cluster.yaml [hunk 11]
```diff
@@ -37,6 +37,10 @@ spec:
   skipUpgradeChecks: false
   # Whether or not continue if PGs are not clean during an upgrade
   continueUpgradeAfterChecksEvenIfNotHealthy: false
+  # WaitTimeoutForHealthyOSD defines the time (in minutes) the operator would wait before an OSD can be stopped for upgrade or restart.
```
Comment (travisn) [inline cluster/examples/kubernetes/ceph/cluster.yaml:40]: I only saw a const 1 minute being used for ok-to-stop. Did you intend to use this setting there as well?

File: pkg/daemon/ceph/client/upgrade.go (hunk 9 above)
Comment (travisn) [reply pkg/daemon/ceph/client/upgrade.go:231]: @leseb Remind me why we even check for ok-to-continue? Seems like we only need to wait if we're checking for ok-to-stop and update an osd. If we continue as soon as the new pod is running, the next OSD will wait anyway on the ok-to-stop, right?

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/daemon/ceph/client/upgrade.go (hunk 10 above)
Comment (sp98) [reply pkg/daemon/ceph/client/upgrade.go:31]: That's a good point. Keeping retry delay to 10s and calculating maxRetries based on the `waitTimeoutForHealthyOSD` set by user. Now:
retries = waitTimeoutForHealthyOSD(in seconds) / 10
delay = 10s

File: pkg/daemon/ceph/client/upgrade.go (hunk 9 above)
Comment (sp98) [reply pkg/daemon/ceph/client/upgrade.go]: With latest changes in the PR, we having following timeout and retry configurations:
1. ok-to-stop
   - OSD:
         - retries = waitTimeoutForHealthyOSD(in seconds)/10
         - delay = 10 seconds. 
         - For example, if user gives waitTimeoutForHealthyOSD as 10 (600 seconds), then, retry count is 60, delay between 
           each retry is 10 seconds.
           
   - Other daemons:
        - retries = 10
        - delay =  1 minute
        - Total = 10 minutes

2  ok-to-continue
   - OSD:  same as ok-to-stop above
   - MDS: 
      - retries = 10
      - delay = 15 seconds
      - Total = 150 seconds. 
   - ok-to-continue is not configured for any other daemons other than OSD and MDS

File: pkg/daemon/ceph/client/upgrade.go [hunk 12]
```diff
@@ -400,3 +411,14 @@ func osdDoNothing(context *clusterd.Context, clusterInfo *ClusterInfo) bool {
 
 	return false
 }
+
+func getRetryConfig(clusterInfo *ClusterInfo, daemonType string) (int, time.Duration) {
+	switch daemonType {
+	case "osd":
```
Comment (leseb) [inline pkg/daemon/ceph/client/upgrade.go:417]:
```suggestion
	case cephv1.KeyOSD:
```

File: pkg/daemon/ceph/client/upgrade.go [hunk 13]
```diff
@@ -400,3 +411,14 @@ func osdDoNothing(context *clusterd.Context, clusterInfo *ClusterInfo) bool {
...
+		return int(clusterInfo.OsdUpgradeTimeout / defaultOSDRetryDelay), defaultOSDRetryDelay
+	case "mds":
```
Comment (leseb) [inline pkg/daemon/ceph/client/upgrade.go:419]:
```suggestion
	case cephv1.KeyMDS:
```

Comment (leseb) [reply pkg/daemon/ceph/client/upgrade.go:419]: You need to add it.

File: pkg/daemon/ceph/client/upgrade.go [hunk 14]
```diff
@@ -400,3 +411,14 @@ func osdDoNothing(context *clusterd.Context, clusterInfo *ClusterInfo) bool {
 
 	return false
 }
+
+func getRetryConfig(clusterInfo *ClusterInfo, daemonType string) (int, time.Duration) {
```
Comment (leseb) [inline pkg/daemon/ceph/client/upgrade.go:415]: unit tests?

File: pkg/daemon/ceph/client/upgrade.go (hunk 12 above)
Comment (leseb) [reply pkg/daemon/ceph/client/upgrade.go:417]: And use `rook.KeyType` instead of a string for `daemonType`

**Review by leseb — CHANGES_REQUESTED** [review]

File: pkg/daemon/ceph/client/upgrade.go (hunk 14 above)
Comment (sp98) [reply pkg/daemon/ceph/client/upgrade.go:393]: added unit tests.

File: pkg/daemon/ceph/client/upgrade.go (hunk 13 above)
Comment (sp98) [reply pkg/daemon/ceph/client/upgrade.go:397]: Like discussed yesterday, not adding keyType constants in this PR as it would requiring changing the `UpdateCephDeploymentAndWait` signature in a lot of places.

File: pkg/daemon/ceph/client/upgrade.go (hunk 9 above)
Comment (sp98) [reply pkg/daemon/ceph/client/upgrade.go]: Removed `ok-to-continue` logic for OSDs

File: cluster/examples/kubernetes/ceph/cluster.yaml [hunk 15]
```diff
@@ -37,6 +37,10 @@ spec:
...
+  # It also defines the time the rook will wait for PGs to become active+clean after an OSD daemon is upgraded or restarted.
```
Comment (leseb) [inline cluster/examples/kubernetes/ceph/cluster.yaml:41]: This is not so true anymore right?

File: cluster/examples/kubernetes/ceph/cluster.yaml [hunk 16]
```diff
@@ -37,6 +37,10 @@ spec:
...
+  # The default wait timeout is 10 minutes.
```
Comment (leseb) [inline cluster/examples/kubernetes/ceph/cluster.yaml:42]: It would be nice to explain what's happening after the timeout.

File: pkg/apis/ceph.rook.io/v1/types.go [hunk 17]
```diff
@@ -98,6 +98,11 @@ type ClusterSpec struct {
 	// ContinueUpgradeAfterChecksEvenIfNotHealthy defines if an upgrade should continue even if PGs are not clean
 	ContinueUpgradeAfterChecksEvenIfNotHealthy bool `json:"continueUpgradeAfterChecksEvenIfNotHealthy,omitempty"`
 
+	// WaitTimeoutForHealthyOSD defines the time the operator would wait before an OSD can be stopped for upgrade or restart.
+	// It also defines the time the rook will wait for PGs to become active+clean after an OSD daemon is upgraded or restarted.
```
Comment (leseb) [inline pkg/apis/ceph.rook.io/v1/types.go:102]: ditto

**Review by leseb — CHANGES_REQUESTED** [review]

File: pkg/daemon/ceph/client/upgrade.go [hunk 18]
```diff
@@ -131,6 +137,7 @@ func EnableReleaseOSDFunctionality(context *clusterd.Context, clusterInfo *Clust
 
 // OkToStop determines if it's ok to stop an upgrade
 func OkToStop(context *clusterd.Context, clusterInfo *ClusterInfo, deployment, daemonType, daemonName string) error {
+	okToStopRetries, okToStopDelay := getRetryConfig(clusterInfo, daemonType)
```
Comment (travisn) [inline pkg/daemon/ceph/client/upgrade.go:140]: Can you move this line inside the switch statement? Or will it be used for daemons other than OSDs?

File: pkg/daemon/ceph/client/upgrade.go [hunk 19]
```diff
@@ -177,14 +187,6 @@ func OkToStop(context *clusterd.Context, clusterInfo *ClusterInfo, deployment, d
...
-	case "osd":
-		if osdDoNothing(context, clusterInfo) {
-			return nil
-		}
-		err := okToContinueOSDDaemon(context, clusterInfo)
-		if err != nil {
-			return errors.Wrapf(err, "failed to check if %s was ok to continue", deployment)
-		}
 	case "mds":
 		err := okToContinueMDSDaemon(context, clusterInfo, deployment, daemonType, daemonName)
```
Comment (travisn) [inline pkg/daemon/ceph/client/upgrade.go:191]: @BlaineEXE @leseb Is there still a use for ok-to-continue with any of the daemons? Or does MDS still need it for some reason? Removing ok-to-continue completely could be a separate PR though...

File: pkg/operator/ceph/cluster/osd/osd.go [hunk 20]
```diff
@@ -61,10 +62,11 @@ const (
...
 	// FailureDomainKey is the label key whose value is the failure domain of the OSD
-	FailureDomainKey  = "failure-domain"
-	prepareAppName    = "rook-ceph-osd-prepare"
-	prepareAppNameFmt = "rook-ceph-osd-prepare-%s"
-	osdAppNameFmt     = "rook-ceph-osd-%d"
+	FailureDomainKey                = "failure-domain"
+	prepareAppName                  = "rook-ceph-osd-prepare"
+	prepareAppNameFmt               = "rook-ceph-osd-prepare-%s"
+	osdAppNameFmt                   = "rook-ceph-osd-%d"
+	defaultWaitTimeoutForHealthyOSD = time.Second * 600
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/osd.go:69]:
```suggestion
	defaultWaitTimeoutForHealthyOSD = 10 * time.Minute
```

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/daemon/ceph/client/upgrade.go (hunk 18 above)
Comment (sp98) [reply pkg/daemon/ceph/client/upgrade.go:140]: This will be used by all the daemons going through `ok-to-stop`. So kept outside the switch statement.  The `getRetryConfig` method returns the default values in cases other than `OSD` and `MDS` daemons.

File: pkg/operator/ceph/cluster/osd/osd.go (hunk 8 above)
Comment (sp98) [reply pkg/operator/ceph/cluster/osd/osd.go:187]: 1. The `UpdateCephDeploymentAndWait` method is used in a lot of places including the nfs operator. If `c.spec` is passed as an argument, then its only purpose would be to use the new `WaitTimeoutForHealthyOSD`. Daemons other than OSD, don't need this new field.
2. We already use 7 arguments in `UpdateCephDeploymentAndWait` 

I was able to make the changes and use `c.spec` as argument and test it. But didn't look like a better approach than using the clusterInfo. Thoughts?

File: cluster/examples/kubernetes/ceph/cluster.yaml (hunk 15 above)
Comment (sp98) [reply cluster/examples/kubernetes/ceph/cluster.yaml]: Yes. Removed the message and added text about what will happen if after timeout

File: pkg/operator/ceph/cluster/osd/osd.go (hunk 20 above)
Comment (sp98) [reply pkg/operator/ceph/cluster/osd/osd.go]: updated to used time in minutes.

File: pkg/daemon/ceph/client/upgrade.go (hunk 19 above)
Comment (BlaineEXE) [reply pkg/daemon/ceph/client/upgrade.go:191]: I haven't looked at MDS stuff in a while. Looks like Seb was the one who created `okToContinueMDSDaemon`. Maybe he remembers something more about it?

Comment (leseb) [reply pkg/daemon/ceph/client/upgrade.go:191]: The fs is much more critical in terms of availability, waiting for the MDS to become active again should happy pretty quickly. I think it's worth waiting.

Comment (travisn) [reply pkg/daemon/ceph/client/upgrade.go:191]: What is there to wait for?  Is there something left in the reconcile?

Comment (leseb) [reply pkg/daemon/ceph/client/upgrade.go:191]: Wait for the MDS to be active again or in standby-replay.

**Review by travisn — APPROVED** [review]
@sp98 LGTM, how about you just rebase to make sure the tests are run against the latest merges?

File: cluster/charts/rook-ceph/templates/resources.yaml [hunk 21]
```diff
@@ -63,6 +63,8 @@ spec:
                   type: boolean
                 continueUpgradeAfterChecksEvenIfNotHealthy:
                   type: boolean
+                waitTimeoutForHealthyOSD:
```
Comment (BlaineEXE) [inline cluster/charts/rook-ceph/templates/resources.yaml:66]: Why not change this to either
- a string which can take `1m` or similar values, OR
- rename this to `waitTimeoutForHealthyOSDInMinutes`
so that users don't have to look at docs to understand units?

**Review by BlaineEXE — CHANGES_REQUESTED** [review]

File: cluster/charts/rook-ceph/templates/resources.yaml (hunk 21 above)
Comment (sp98) [reply cluster/charts/rook-ceph/templates/resources.yaml]: using `waitTimeoutForHealthyOSDInMinutes`.

**Comment by sp98** [comment]
> @sp98 LGTM, how about you just rebase to make sure the tests are run against the latest merges?

rebased with latest merges

**Review by leseb — APPROVED** [review]

**Review by BlaineEXE — APPROVED** [review]

**Comment by jshen28** [comment]
I am thinking `ok-to-continue` is necessary because we might not want to have a high recovery bandwidth. Imo `ok-to-stop` does not check that.
